### PR TITLE
#184347029 mark product as available or not

### PR DESCRIPTION
**Pull request by Fidela1**
#### What does this PR do?

1. Allow Seller to mark  product as available or not
2. Allow seller to update and delete product

#### Description of Task to be completed?

- [x] Mark  product as available or not
- [x] Delete product
- [x] Update product

#### How should this be manually tested?

1. clone [repo](url)
2. Change into the directory vikings-ec-bn
3. Checkout to this branch `feature/available-product-184347029`
4. Run command `npm install`
5. Run `npm  run dev`
6. Open browser and visit http://localhost:5000/api-docs/#/
7. Test functionalities of product/ update product, delete product and mark available product

#### What are the relevant pivotal tracker stories?

[184347029](https://www.pivotaltracker.com/story/show/184347029)
[184347030](https://www.pivotaltracker.com/story/show/184347030)
[184347033](https://www.pivotaltracker.com/story/show/184347033)

### CODE REVIEW

File: src/controllers/product.controller.js [hunk 1]
```diff
@@ -37,3 +39,38 @@ export const getAllProducts = async (req, res) => {
       .json({ error: err.message, message: 'Failed to retrieve products' });
   }
 };
+
+export const markAvailableProduct = async (req, res) => {
+  try {
+    const isAvailable = req.body.isAvailable;
```
Comment (TuyizeeAnastase) [inline src/controllers/product.controller.js]: use destructuring assignment

File: src/controllers/product.controller.js [hunk 2]
```diff
@@ -37,3 +39,38 @@ export const getAllProducts = async (req, res) => {
...
+    await ProductService.markProduct({ isAvailable }, req.params.productId);
+    return res.status(200).json({ message: 'product marked' });
```
Comment (TuyizeeAnastase) [inline src/controllers/product.controller.js]: change message `Product is available now`

File: src/controllers/product.controller.js [hunk 3]
```diff
@@ -37,3 +39,38 @@ export const getAllProducts = async (req, res) => {
...
+  } catch (error) {
+    return res.status(500).json({
+      error: error.message,
+      message: 'Failed to mark product',
+    });
+  }
+};
+export const updateProduct = async (req, res) => {
+  try{
+    console.log('------------',req.body,req.params.productId);
```
Comment (TuyizeeAnastase) [inline src/controllers/product.controller.js]: Don't push console

File: src/routes/api/product.routes.js [hunk 4]
```diff
@@ -28,4 +32,28 @@ productRoutes.post(
...
+  productValidation,
+  checkIfProductExistById,
+  updateProduct,
+);
+productRoutes.put(
+  '/:productId',
+  protectRoute,
+  restrictTo('seller','admin'),
+ validateAvailable,
+ checkIfProductExistById,
```
Comment (TuyizeeAnastase) [inline src/routes/api/product.routes.js]: add middleware to check if the product is not available before made Available.

File: src/services/product.service.js [hunk 5]
```diff
@@ -31,4 +31,17 @@ export class ProductService {
...
+  static async markProduct(fields, id) {
+    return await Products.update({ ...fields }, { where: { id: id } });
+  }
+
+static async updateProduct(fields, id)  {
+  return await Products.update({ ...fields }, { where: { id: id } });
+}
+ 
+static async deleteProduct(id){
+  console.log('-------',id);
```
Comment (TuyizeeAnastase) [inline src/services/product.service.js]: don't push debugging tools like console

**Review by TuyizeeAnastase — CHANGES_REQUESTED** [review]

File: src/controllers/product.controller.js [hunk 6]
```diff
@@ -37,3 +39,42 @@ export const getAllProducts = async (req, res) => {
...
+    {return res.status(200).json({ message: 'product is available now' });}
+    else{
+      return res.status(200).json({ message: 'product is not available now' });
+    }
+  } catch (error) {
+    return res.status(500).json({
+      error: error.message,
+      message: 'Failed to mark product',
+    });
+  }
```
Comment (ncutixavier) [inline src/controllers/product.controller.js]: - We need to have a global service to update products, and that's what you'll be using to update product availability
- You're just updating, why are you checking availability?
- Write a meaningful message in the catch

File: src/controllers/product.controller.js [hunk 7]
```diff
@@ -37,3 +39,42 @@ export const getAllProducts = async (req, res) => {
...
+};
+export const updateProduct = async (req, res) => {
+  try{
+   
+     await ProductService.updateProduct(req.body, req.params.productId);
```
Comment (ncutixavier) [inline src/controllers/product.controller.js]: Use this service to update product availability (Marking product)

File: src/middlewares/product.middleware.js [hunk 8]
```diff
@@ -10,7 +10,27 @@ export const checkIfProductExists = async (req, res, next) => {
...
+  return res
+			.status(404)
+			.json({ message: 'Product is already available' });
+}
+else if (product.isAvailable === req.body.isAvailable && req.body.isAvailable === false)
+return res
+			.status(404)
+			.json({ message: 'Product is already not available' });
+	req.product = product;
+	next();
```
Comment (ncutixavier) [inline src/middlewares/product.middleware.js]: Let's have separate middleware

File: src/services/product.service.js [hunk 9]
```diff
@@ -31,4 +31,16 @@ export class ProductService {
     const totalItems = count;
     return { totalPages, currentPage, totalItems, rows };
   }
+
+  static async markProduct(fields, id) {
+    return await Products.update({ ...fields }, { where: { id: id } });
+  }
```
Comment (ncutixavier) [inline src/services/product.service.js]: We don't need this service, since you're already have the one to update product

**Review by ncutixavier — CHANGES_REQUESTED** [review]

**Review by irakozetony — APPROVED** [review]

**Review by Abdden — APPROVED** [review]

**Review by fabmukunzi — APPROVED** [review]

File: src/database/config/config.js [hunk 10]
```diff
@@ -2,18 +2,21 @@ require('dotenv').config();
...
     dialect: 'postgres',
     dialectOptions: {
-      ssl: {
-        require: true,
-        rejectUnauthorized: false,
-      },
+      // ssl: {
+      //   require: true,
+      //   rejectUnauthorized: false,
+      // },
```
Comment (ncutixavier) [inline src/database/config/config.js]: Remove all unnecessary comments

File: src/routes/api/product.routes.js [hunk 11]
```diff
@@ -59,4 +62,27 @@ productRoutes.get(
...
+  checkIfProductIsAvailableById,
+  markAvailableProduct,
+);
+
+   productRoutes.delete(
+    '/:productId',
+    protectRoute,
+    restrictTo('seller', 'admin'),
+    checkIfProductExistsById,
+    deleteProduct
```
Comment (ncutixavier) [inline src/routes/api/product.routes.js]: You're missing a middleware to check if the products belong to the seller, and it's already implemented

File: src/routes/api/product.routes.js [hunk 12]
```diff
@@ -59,4 +62,27 @@ productRoutes.get(
...
+  checkIfProductExistsById,
+  updateProduct
+);
+
+productRoutes.put(
+  '/:productId',
+  protectRoute,
+  restrictTo('seller','admin'),
+  checkIfProductIsAvailableById,
+  markAvailableProduct,
```
Comment (ncutixavier) [inline src/routes/api/product.routes.js]: You're missing validation and middleware to check the owner

File: src/routes/api/product.routes.js [hunk 13]
```diff
@@ -59,4 +62,27 @@ productRoutes.get(
   checkIfSellerOwnsProduct,
   getSpecificProduct
 );
+productRoutes.patch('/:productId',
+  protectRoute,
+  restrictTo('seller', 'admin'),
+  productValidation,
+  checkIfProductExistsById,
+  updateProduct
```
Comment (ncutixavier) [inline src/routes/api/product.routes.js]: Missing middleware to check the owner

File: src/controllers/product.controller.js [hunk 14]
```diff
@@ -108,6 +110,41 @@ export const getSpecificProduct = async (req, res) => {
...
       error: err.message,
       message: 'Error occured while retrieving a product',
+      
+    });
+  }
+};
+export const markAvailableProduct = async (req, res) => {
+  try {
+    const {isAvailable} = req.body;
+    await ProductService.updateProduct( isAvailable, req.params.productId);
```
Comment (ncutixavier) [inline src/controllers/product.controller.js]: - You don't have to use `req.params`, when you have a middleware that returns the product
- Validation is needed

File: src/controllers/product.controller.js [hunk 15]
```diff
@@ -108,6 +110,41 @@ export const getSpecificProduct = async (req, res) => {
...
+  } catch (error) {
+    return res.status(500).json({
+      error: error.message,
+      message: 'Failed to mark if product is available or not',
     });
   }
 };
+export const updateProduct = async (req, res) => {
+  try{
+     await ProductService.updateProduct(req.body, req.params.productId);
```
Comment (ncutixavier) [inline src/controllers/product.controller.js]: You don't have to use req.params, when you have a middleware that returns the product

File: src/controllers/product.controller.js [hunk 16]
```diff
@@ -108,6 +110,41 @@ export const getSpecificProduct = async (req, res) => {
...
+    .json({ message: 'Product updated successfuly' });
+  } catch (err) {
+    res.status(500).json({ error: err.message,
+       message: 'fail to update product'});
+  }
+};
+ 
+export const deleteProduct = async (req, res) => {
+  try {
+     await ProductService.deleteProduct(req.params.productId);
```
Comment (ncutixavier) [inline src/controllers/product.controller.js]: You don't have to use req.params, when you have a middleware that returns the product

File: src/controllers/product.controller.js [hunk 17]
```diff
@@ -108,6 +110,41 @@ export const getSpecificProduct = async (req, res) => {
...
   }
 };
+export const updateProduct = async (req, res) => {
+  try{
+     await ProductService.updateProduct(req.body, req.params.productId);
+    return res.status(200)
+    .json({ message: 'Product updated successfuly' });
+  } catch (err) {
+    res.status(500).json({ error: err.message,
+       message: 'fail to update product'});
```
Comment (ncutixavier) [inline src/controllers/product.controller.js]: Rephrase this message

**Review by ncutixavier — CHANGES_REQUESTED** [review]

**Review by ncutixavier — APPROVED** [review]

File: src/database/config/config.js [hunk 18]
```diff
@@ -14,6 +14,7 @@ module.exports = {
   },
   test: {
     url: process.env.TEST_DATABASE_URL,
+    //  url: process.env.FIDELA_REMOTE,
```
Comment (ncutixavier) [inline src/database/config/config.js]: Remove unnecessary comments

**Review by ncutixavier — APPROVED** [review]

**Pull request merged**